### DISCUSSION TITLE
CMS

### DISCUSSION THREAD

**Discussion by andrecasal** [epicweb-dev/epic-stack#160] [Ideas]
Shall we start fleshing out a proper CMS?

I'm thinking admin, manager, and content writer account management, with a create -> review -> publish production cycle.

I'd be happy to open an initial PR for this.

**Comment by kentcdodds** [comment]
Thanks for offering to do that! Normally I'd be happy to have you start work on this. But this is one thing I really want to do myself. Maybe you can make [an example](https://github.com/epicweb-dev/epic-stack/blob/main/docs/examples.md) for folks to reference as an idea of how to do this?

**Reply by lucascurti** [reply]
@kentcdodds Are you thinking about using a 100% custom implementation for the CMS, or you'll use some library as a base? Asking because I need to add a CMS to a project and, if you are thinking about using a library for it, I would choose the same one so migrating it to your implementation, once ready, won't be as hard.

**Reply by kentcdodds** [reply]
I'm not certain. Part of me feels like going for a custom one is naively optimistic about how easy it will be. But I rarely am satisfied with the level of complexity offered by libraries for stuff like this.

**Reply by kentcdodds** [reply]
At this point, I certainly couldn't tell you which library I would use if we do end up with a library though.

**Reply by mwheatfill** [reply]
@kentcdodds not sure how far along you are with this, but an interesting library to keep your eye on is [keystatic](keystatic.com). Content is stored locally and/or Github (no additional service required), supports TypeScript, and has a Remix integration guide.

**Reply by kentcdodds** [reply]
@simonswiss would probably be interested in putting together a keystatic + Epic Stack example 😉 https://github.com/epicweb-dev/epic-stack/blob/main/docs/examples.md

**Reply by mwheatfill** [reply]
Yeah, he probably knows a thing or two about it! 😁

**Reply by simonswiss** [reply]
Well yes I'd sure be interested in doing that :)

The only thing to know about Keystatic is it's mostly going to serve read-only scenarios — there are no CRUD operations enabled on the frontend with it, it just makes almost any bit of JSON/YAML/Markdoc/MDX editable via an Admin UI.

Also, it won't sync with the existing sqlite DB — but will read/write directly on local files. The "database" in the world of Keystatic is a collection of JSON/YAML/Markdoc/MDX files, that can sit anywhere on the project tree.

In other words, it would be perfect to manage a blog, marketing pages, documentation etc.. but not to create/update/delete notes from the existing CRUD Notes app.

So if the hope was to have a CMS that seamlessly hops onto managing content in the context of the demo notes app, this would not be it.

However, if the idea is to have a blog demo part of an epic stack starter, this would work beautifully.

If you want CRUD + a real DB (which can be sqlite), [Keystatic's older brother Keystone JS](https://github.com/epicweb-dev/epic-stack/discussions/160#discussioncomment-7089600) (built by the same team than Keystatic) is an excellent choice.

**Reply by kentcdodds** [reply]
I think that's a great thing to call out. Adding a `/blog` portion to the epic stack would be a great way to demonstrate how to use keystatic to help with the kinds of problems it's intended for :)

**Reply by simonswiss** [reply]
Sound great — onto it!

**Reply by simonswiss** [reply]
Soooo... here's a little [video demo](https://www.dropbox.com/scl/fi/hzsjvqrb8ir90lo3yw9f9/Epic-Stack-Keystatic.mp4?rlkey=u5hfbant07l6klnanqhy0o3xf&dl=0) of the progress update :)

There is one major painpoint I am facing: Keystatic wants to become part of your Remix app and render routes (and API routes) via the `app/routes` tree.

BUT — The Epic Stack has a pretty significant `root.tsx` file, and all of that (the logic, the header/footer) gets passed to the Keystatic Admin UI routes.

The solution would be to move the complexity of the `root.tsx` file to a `layout` file somewhere, and use nested layouts to wrap all the app (except Keystatic) with it, but... this would add *a lot of noise* to the PR 😅

I wish there was a way to opt out of the `root.tsx` file and provide an alternate entry point for a given route, when you want it to have nothing to do with the rest of the app.

In Next.js, this can be bypassed with the concept of "grouped routes", which will take their own layout independently of the rest of the app. Is there a way to do such a thing without moving anything from the current `root.tsx`?

You can see the work in progress in [this repo](https://github.com/simonswiss/epic-stack-with-keystatic).

![CleanShot 2024-03-05 at 13 44 26@2x](https://github.com/epicweb-dev/epic-stack/assets/485747/108e05a0-e710-43dd-9035-8a95276f7024)

**Reply by simonswiss** [reply]
Here's a commit that tells a [pretty compelling story](https://github.com/simonswiss/epic-stack-with-keystatic/commit/a5fb10e76311544fbc18d32775cdd1210910683b).

It adds an example "YouTube video with caption" custom MDX block that can be slotted anywhere inside any blog post.

The same UI component is used on the front end and in the Admin UI for the `ContentView` preview 🤙

![CleanShot 2024-03-05 at 14 24 57@2x](https://github.com/epicweb-dev/epic-stack/assets/485747/300dedce-a7ab-4fb7-a241-972d6dee3f1a)

![CleanShot 2024-03-05 at 14 27 21@2x](https://github.com/epicweb-dev/epic-stack/assets/485747/a2ca732b-b2e8-4f21-81dc-2f89b2870a14)

![CleanShot 2024-03-05 at 14 28 01@2x](https://github.com/epicweb-dev/epic-stack/assets/485747/4f67e3d9-0c66-4891-80a3-c8f6fb61191c)

**Reply by mwheatfill** [reply]
Wow, this looks awesome @simonswiss! Regarding the issue of root.tsx rendering for the admin UI, @kentcdodds  isn't there a way to not render the parent route as an option? I think you do that with your blog posts and remember you mentioning something about it.

**Reply by simonswiss** [reply]
If there is, that would be awesome!

**Reply by kentcdodds** [reply]
Yes there is. You can learn about the filename convention here: https://github.com/kiliman/remix-flat-routes#conventions

So for this, you could do `app/routes/admin_.tsx` or if you need multiple routes you'd do: `app/routes/admin_.tsx` (`/admin`) and `app/routes/admin_.content.tsx` (`/admin/content`) or if you wanted a directory of routes under that path you could do `app/routes/admin_+/index.tsx` and `app/routes/admin_+/content.tsx`.

Pretty versital. But the key to ignoring the parent layout is the trailing underscore.

**Reply by simonswiss** [reply]
Amazing! I read those docs and also asked in Remix Discord but somehow completely missed that part. Great stuff! 🎉

**Reply by simonswiss** [reply]
Hmmm, looks like these (with trailing underscore) will fallback to `root.tsx`?

![CleanShot 2024-03-06 at 06 27 39@2x](https://github.com/epicweb-dev/epic-stack/assets/485747/8c2f613e-7e44-4fcd-9697-689f4fdf9625)

But what if `_root.tsx` is the thing I'm trying to avoid (and replace with a barebones Remix <html> scaffolding)?

Essentially, I'd like the `routes/keystatic_+/$.tsx` to use the "vanilla" Remix root:

```tsx
import {
  Links,
  Meta,
  Outlet,
  Scripts,
  ScrollRestoration,
} from "@remix-run/react";

export default function KeystaticRoot() {
  return (
    <html lang="en">
      <head>
        <Links />
        <Meta />
      </head>
      <body>
        <Outlet />
        <ScrollRestoration />
        <Scripts />
      </body>
    </html>
  );
}
```

**Reply by kentcdodds** [reply]
Ahh, yes, you can't avoid what's in the `root.tsx` being a wrapper around your app. So what you'll want to do is add a layout route for what's currently in the app (basically take all the stuff in `app/routes` and put it in `app/routes/_app+`) and then move as much as possible from `app/root.tsx` to `app/routes/_app+/_layout.tsx`, then you'll be able to do what I suggested before with much less in the way from the rest of the app.

Sorry it's kind of confusing. This is one reason the Remix team wants to get rid of file-based routing. It's like a whole other language. For myself I have mixed feelings because I think once learned it's valuable, but until you decide it's a thing to be learned it's just really annoying.

**Reply by kiliman** [reply]
Another option is to change your _root_ to render different layouts based on the route. This way, you don't have to change your existing route filenames.

There are a couple of ways to do this:

1. Simply `useLocation`, check the path, and render a different `<Layout>` for `/blog/*`, for example.
2. Another option is to export a custom layout as a `handle` in your `blog` route. Then, in _root_, it looks for the layout handle via `useMatches`.

```ts
// root.tsx

function DefaultLayout() {
  // whatever your app layout would be
}

export default function App() {
  // see if there's a route that has a custom Layout handle
  const Layout = useMatches().find(m => m.handle?.Layout)?.handle.Layout ?? DefaultLayout

  return <Layout><Outlet /></Layout>
}

// routes/blog.tsx
export const handle = { 
  Layout: CustomLayout
}

function CustomLayout() {
  return // custom layout
}
```

**Reply by mwheatfill** [reply]
@simonswiss sorry I thought there may have been a way to avoid such large changes to root.tsx! Thanks for the explanation @kentcdodds

**Reply by simonswiss** [reply]
I did something similar to what @kiliman suggests, using `useLocation()` to determine if this is a Keystatic route, and render a much simpler layout if that's the case.

Still had to pass the `nonce` for it to work properly — not sure if this is the ideal setup but it works great!

https://github.com/simonswiss/epic-stack-with-keystatic/commit/3959a0ab9dfed740c13eff7c0d00140b1f436d00

**Comment by vinstah** [comment]
since Strapi is open source that might be a good place to look into seeing how some parts can work in Epic Stack

**Reply by cr101** [reply]
WordPress is another popular open source option

**Reply by vinstah** [reply]
yeah, Strapi has the advantage of being JS/React not PHP, similar though is SilverStripe CMS https://silverstripe.github.io/silverstripe-pattern-lib/

**Reply by supersonicart** [reply]
Is WordPress used that often with React based frameworks?

**Reply by victorkane** [reply]
WordPress _is_ a React based framework. It uses the new block based editor which is something that I believe is a React SSG. I mean, you click on Save, and React blocks are compiled into WordPress-ese (html + block data expressed in html comments) which is stored in the database as content. Visitors get that. It's a SSG and uses the Gutenberg (now called "WordPress") editor. It's worth a long conversation!!!

**Reply by victorkane** [reply]
I have spent many years with Drupal which, while WordPress eschews "custom post types", revels in structured content and separation of concerns in a quasi-MVC context. Unlike Directus (which made the switch from PHP to Node/Vue and so can scale and is awesome except even with a paid cloud plan you don't get access to the database you're using), it is (my opinion) in slow and not so slow decline since it cannot scale even though it keeps "renovating" itself on an over-complexity island. But many of its principles (hooks, separation of concerns, importance of structured content and built-in visual data query builder, to name a few) are worthy of discussion. Another CMS that made the switch from PHP to Node.js is Keystone, which I also tried to use (another worthy discussion). Looking forward to discussing and sharing my experience using any of the CMS I am mentioning; yes CMS's have played a large part in my life...

**Reply by victorkane** [reply]
I have used Strapi for several years (early on in its existence) and it is worthy of discussion because it has remained open source, is built on full stack principles; and for the (my opinion) huge mistakes it has made and has built-in, unfortunately. One fantastic example is that instead of using Prisma, it directly supports SQLite, MongoDB and at least one SQL server driven database (can't remember if it's MySql or PostgreSQL (one is implemented, the other is "coming soon")) instead of using an ORM (as Remix brilliantly does with Prisma). So Strapi is worthy of discussion because 1) It's open stack and has great strengths; and 2) It's baked-in flaws which (my opinion) stunt its growth.

**Comment by supersonicart** [comment]
I am using Sanity with my Epic Stack for basic content management and blog posts: it is working epically well!

**Reply by andrecasal** [reply]
I love Sanity, but even though you can self-host its frontend, you can't self-host your data. This creates a third-party service dependency which isn't ideal. The content is also centralized in Sanity's Content Lake, in US servers, which raises some data privacy issues for folks in the EU.

**Reply by supersonicart** [reply]
Sanity hosting the content really simplifies things for me; but then again a CMS built into the Epic Stack framework would be much nicer.  Especially if it had features like user analytics, stripe integration/analytics, etc.

**Reply by andrecasal** [reply]
Giving Sanity a try. The `post` schema is already defined in `/sanity/schemas/post.ts`, but it seems Sanity doesn't have a way to convert this into a type, so there's no type safety across the network boundary when we query for posts (or anything really). I'm using zod to validate the data coming from the queries, which is redundant. How did you solve this? 😅

**Reply by supersonicart** [reply]
I'd suggest take a look in Sanity's docs - or hitting them up on their Slack for the directly specific answer you're looking for.  I believe the answer lies in the way you query your data?

**Comment by andrecasal** [comment]
I tried out Sanity on a branch and here are some cons to consider:
- 3rd party dependency
- Data lives on Sanity servers
- Requires you to mock this service for local dev and testing
- Pulling data in requires you to learn [GROQ](https://www.sanity.io/docs/groq) (it's easy)
- Sanity gives you article/blog data formatted using [Portable Text](https://github.com/portabletext/portabletext), which adds [another package](https://github.com/portabletext/react-portabletext) to the frontend, doesn't add any functionality over Markdown, and takes some work to extend if you have custom elements like `<CodeBlock>`'s, for example.

The only 'pro' I found is that the DX is excellent with Sanity Studio.

Sanity's use-case is when you want to create content at scale and distribute it to multiple channels.

That being said, I'm going back to what I've always used: MDX. The only con is that each article is its own file. But in practice, this just means that after a few tens of articles you start using search on the file system (just like you would use search on your CMS's backend). It's not a big drawback and you have the full power of JSX to insert custom components (code blocks, design elements, backgrounds, etc) in the middle of your articles.

### Update

After tinkering with MDX for a few hours, given you have a UI library to lean on, I saw little benefit in writing time and plenty of formatting drawbacks.

P.S.: I'm open sourcing [my own UI library](https://andrecasal.com/ui) soon, if you're interested.

**Reply by victorkane** [reply]
For MDX files in the file system, the flat file CMS [Front Matter CMS](https://frontmatter.codes/) (a VS Code extension) _may_ be a solution, with images, categories and tags also, as well as relationsips between articles, authors, etc. which it manages via schema. They mention Hugo, Next.js and Astro integration in the docs, but the configuration is customizable. It offers another interesting approach to study.

**Comment by victorkane** [comment]
Hi everyone. Sorry I'm making so many comments, but CMS is really my thing. Got one question, tho... Isn't the Epic Stack its own CMS, in many ways? Or, could be. I do understand that the goal of the CMS would be to improve team synergy via a single source of truth for two things: structured content modeling for each project, and the content creation workflow. Then we need a Figma for single source of truth for design system. WordPress, for example, tries to be all things for all persons, which is a strength and a weakness (or is it?). OK, so in this discussion "CMS" really needs to be defined.

**Comment by victorkane** [comment]
One reference example must-see I believe is [Payload CMS](https://payloadcms.com/) "headless" CMS, schema based, depending on Node.js and MongoDB. Write typescript config schema -> 1) DB created 2) (customizable) Admin UI created 3) API created -> You manage your own server Express app 

> No black magic, all TypeScript, and fully open-source.

They have both paid cloud as well as self-hosting.

- Docs: https://payloadcms.com/docs/getting-started/what-is-payload
- Demo: https://demo.payloadcms.com/admin/create-first-user
- Recently streamed discussion on YouTube concerning their immediate and future roadmap: [Payload Planning: Live Preview and Improved Admin UI](https://youtu.be/7OCeE84rRBM?si=1ZADQU_PV9_b8tEp)
- GitHub discussion on Roadmap: https://github.com/payloadcms/payload/discussions/categories/roadmap
  - Prisma for more db???
  - Switching from Slate to Lexical markdown editor component
  - Lot of consideration on UI and accessibility for technical as well as non-technical users
- Their architecture seems to be much leaner and meaner than Strapi but with obvious similarities

**Reply by MoSattler** [reply]
Payload is great, I got to use it a couple of times, and I really like it. 
Looks like SQLite support [will be shipped soon](https://github.com/payloadcms/payload/discussions/287#discussioncomment-7944818).

The only problem I see is that they use Drizzle, while Epic-Stack uses Prisma. 

@kentcdodds Would love to hear your thoughts on PayloadCMS.

**Comment by bjornbrasse** [comment]
What about [KeystoneJS](https://keystonejs.com/), headless CMS...??
